### PR TITLE
Add Recommendations Toggle

### LINKED ISSUE #72: Show / Hide Recommendations View
I'd like the ability to show / hide recommendations.

Example, when I switch to grid view I find the Recommendations visually distracting.

- [ ] Requires https://github.com/nextcloud/server/issues/16602

### LINKED ISSUE #108: Recommendations toggle in bottom left Files settings
**Is your feature request related to a problem? Please describe.**

Discussed in nextcloud/recommendations#100 with @jancborchardt. 

We are currently working on adding the ability for users to disable Recommendations for themselves. The previous plan was to have a separate settings page for Recommendations but  it would be nice to add the setting to the sidebar of the Files app instead. The idea would be for the Files app will check if the Recommendations app is installed, then if installed, give the setting to toggle Recommendations.

If the idea sounds okay, I'll open a pull request.

### PR DESCRIPTION
**Pull request by gary-kim**
Use the new `OCA.Files.Settings` API.
- [x] Requires nextcloud/server#18556
- [x] Only download recommendations when needed
    - [x] Do not find recommendations if disabled
    - [x] On first enable of recommendations by a user, fetch recommendations.

Resolves #72
Resolves #108

Signed-off-by: Gary Kim <gary@garykim.dev>

### CODE REVIEW

**Comment by jancborchardt** [comment]
Just a sidenote, when testing this with the server pull request I get:

> The files of the app Recommendations (recommendations) were not replaced correctly. Make sure it is a version compatible with the server.

Cause the max-version is 18 – setting to 19 fixes it. :)

File: src/components/Settings.vue [hunk 1]
```diff
@@ -0,0 +1,32 @@
+<template>
+	<div id="recommendations-setting-enabled">
+		<input id="recommendationsEnabledToggle"
+			v-model="enabled"
+			class="checkbox"
+			checked="checked"
+			type="checkbox"
+			name="enabled">
+		<label for="recommendationsEnabledToggle">{{ t('recommendations', 'Recommendations enabled') }}</label>
```
Comment (jancborchardt) [inline src/components/Settings.vue]: I’d say better would be "Show recommendations" for uniformity. :)
```suggestion
		<label for="recommendationsEnabledToggle">{{ t('recommendations', 'Show recommendations') }}</label>
```

As e.g. if we ever have recommendations on mobile, the settings should be independent and purely dealing with the visibility on that specific client.

Comment (jancborchardt) [reply src/components/Settings.vue]: Besides that it works great! :)

Comment (gary-kim) [reply src/components/Settings.vue]: Done!

File: lib/Controller/SettingsController.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+		$this->userSession = $userSession;
+	}
+
+	/**
+	 * @NoAdminRequired
+	 *
+	 * @return JSONResponse
+	 * @throws Exception
+	 */
+	public function getSettings (): JSONResponse {
```
Comment (skjnldsv) [inline lib/Controller/SettingsController.php]:
```suggestion
	public function getSettings(): JSONResponse {
```

File: lib/Controller/SettingsController.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+	/**
+	 * @NoAdminRequired
+	 *
+	 * @param $key
+	 * @param $value
+	 * @return JSONResponse
+	 * @throws Exception
+	 */
+	public function setSetting (string $key, string $value): JSONResponse {
```
Comment (skjnldsv) [inline lib/Controller/SettingsController.php]:
```suggestion
	public function setSetting(string $key, string $value): JSONResponse {
```

File: lib/Controller/SettingsController.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+class SettingsController extends Controller {
+
+	/** @var IConfig */
+	private $config;
+
+	/** @var IUserSession */
+	private $userSession;
+
+	public function __construct($appName, IRequest $request, IConfig $config, IUserSession $userSession) {
```
Comment (skjnldsv) [inline lib/Controller/SettingsController.php]:
```suggestion
	public function __construct($appName,
								IUserSession $userSession,
								IRequest $request,
								IConfig $config,
								IUserSession $userSession) {
```

File: lib/Controller/SettingsController.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+	 * @NoAdminRequired
+	 *
+	 * @param $key
+	 * @param $value
+	 * @return JSONResponse
+	 * @throws Exception
+	 */
+	public function setSetting (string $key, string $value): JSONResponse {
+		$user = $this->userSession->getUser();
+		if (is_null($user)) {
```
Comment (skjnldsv) [inline lib/Controller/SettingsController.php]:
```suggestion
		if (!$user instanceof IUser) {
```
More precise :)

File: lib/Controller/SettingsController.php [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+	/**
+	 * @NoAdminRequired
+	 *
+	 * @return JSONResponse
+	 * @throws Exception
+	 */
+	public function getSettings (): JSONResponse {
+		$user = $this->userSession->getUser();
+		if (is_null($user)) {
```
Comment (skjnldsv) [inline lib/Controller/SettingsController.php]:
```suggestion
		if (!$user instanceof IUser) {
```

File: lib/Controller/SettingsController.php [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+ */
+
+namespace OCA\Recommendations\Controller;
+use OCA\Recommendations\AppInfo\Application;
+use OCP\AppFramework\Controller;
+use OCP\AppFramework\Http;
+use OCP\AppFramework\Http\JSONResponse;
+use OCP\IConfig;
+use OCP\IRequest;
+use OCP\IUserSession;
```
Comment (skjnldsv) [inline lib/Controller/SettingsController.php]:
```suggestion
use OCP\IUser;
use OCP\IUserSession;
```

File: package.json [hunk 8]
```diff
@@ -19,7 +19,7 @@
   },
   "homepage": "https://github.com/ChristophWurst/files_recommendation#readme",
   "dependencies": {
-    "nextcloud-axios": "^0.2.1",
+    "@nextcloud/axios": "^1.1.0",
```
Comment (skjnldsv) [inline package.json]: :tada:

File: src/main.js [hunk 9]
```diff
@@ -48,9 +51,20 @@ OC.Plugins.register('OCA.Files.FileList', {
 
 		const View = Vue.extend(Recommendations);
 		const vm = new View({
-			propsData: {}
+			propsData: {},
+			store,
 		}).$mount(this.el);
 
+		// register Files App Setting
+		const settingsView = new (Vue.extend(Settings))({
```
Comment (skjnldsv) [inline src/main.js]: I still find this syntax super confusing :(

File: src/store/store.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+			state.enabled = val;
+		},
+	},
+	actions: {
+		enabled(ctx, val) {
+			ctx.commit('enabled', val);
+			Axios.put(generateUrl('apps/recommendations/settings/enabled'), {
+				value: val.toString(),
+			});
+		},
```
Comment (skjnldsv) [inline src/store/store.js]:
```suggestion
		/**
		 * Toggle the recommendations
		 * @param {Object} context the store context
		 * @param {boolean} enabled recommendations status
		 */
		enabled(context, enabled) {
			context.commit('enabled', enabled);
			axios.put(generateUrl('apps/recommendations/settings/enabled'), {
				value: enabled.toString(),
			});
		},
```


Let's be consistent and easy to understand, we're compiling anyway, so bytes is not an issue here :stuck_out_tongue_closed_eyes:

File: src/store/store.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import Vue from 'vue';
+import Vuex from 'vuex';
+import Axios from '@nextcloud/axios';
```
Comment (skjnldsv) [inline src/store/store.js]:
```suggestion
import axios from '@nextcloud/axios';
```

We use it as axios everywhere, any reason you went for PascalCase? :)

Comment (gary-kim) [reply src/store/store.js]: Habit from other projects :), whoops.

**Comment by gary-kim** [comment]
> Code looks good!
> 
> Is this a production build? `+2,259 −861` seems a bit much thinking

It's from `package-lock.json`, this branch has been around for a while so it looks the like rebase got confused. Give me a sec.

**Comment by ChristophWurst** [comment]
@gary-kim rebase
@gary-kim merge :P

**Comment by szaimen** [comment]
Is this also configurable via occ for all users?

**Comment by kosssi** [comment]
@szaimen have you find a solution to disable recommendation by default ?

**Comment by szaimen** [comment]
No, no official solution but there are many workarounds:
- hide the section with the custom_css app
- build a script that loops over all users and sets the correct user-setting to disable it
- disable the recommendations app

**Comment by corpix** [comment]
this PR appeared in google when I searched  for a way to disable recommendations. After reading PR have no idea how should I disable recommendations, probably something changed, so done this a barbarian way via Nginx configuration:

```
location = /ocs/v2.php/apps/recommendations/api/v1/recommendations {
    return 200 "{\"ocs\":{\"meta\":{\"status\":\"ok\",\"statuscode\":200,\"message\":\"OK\"},\"data\":{\"enabled\":false,\"recommendations\":[]}}}";
    default_type application/json;
}
```

would save me some CPU cycles :laughing:

**Pull request merged**